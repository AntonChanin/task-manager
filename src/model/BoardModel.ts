import BaseModel from './BaseModel';
import GroupModel from './GroupModel';

class BoardModel extends BaseModel {
  override __default = {
    ...super.__default,
    name: 'BoardModel',
    classList: {
      group: 'flex max-w-full overflow-x-auto overflow-y-hidden h-[95%]',
      title: 'my-2 w-max text-white text-2xl',
      paper: 'flex-col bg-blue-400 p-4 font-sans w-full h-full',
    },
    items: [
      new GroupModel({
        name: '',
        description: 'init value',
      })
    ],
  };

  override items: GroupModel[] = [];

  constructor(options: Record<string, any>) {
    const { name, description, parent } = options;
    super({ name, description, parent });

    this.classList = this.__default.classList;  
  };

  addGroup = (newGroup: GroupModel) => {
    this.addItem(newGroup);
  }
};

export default BoardModel;
