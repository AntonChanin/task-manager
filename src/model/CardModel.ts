import BaseModel from './BaseModel';
import { Variant } from '../types/common';
import { BaseProps } from '../types/model';

class CardModel extends BaseModel {
  override  __default: BaseProps =  {
    ...super.__default,
    name: 'CardModel',
    classList: {
      editButton: 'opacity-0 flex ml-auto hover:opacity-100 absolute top-0 right-0',
    },
  };

  variant: Variant = 'secondary';

  constructor(options: Record<string, any>) {
    const { name, description, draggable, parent, variant } = options;
    super({ name, description, draggable, parent });

    this.variant = variant;
    this.classList = this.__default.classList;
  };
};

export default CardModel;
