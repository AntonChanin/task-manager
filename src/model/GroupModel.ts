import BaseModel from './BaseModel';
import CardModel from './CardModel';

class GroupModel extends BaseModel {
  override __default = {
    ...super.__default,
    name: 'GroupModel',
    classList: {
      button: 'flex',
      cards: 'overflow-y-auto max-h-[78vh] pr-2',
      title: 'flex place-content-between py-4',
      paper: 'w-1/4 min-w-[25%] h-max px-4 mr-2 flex-col items-start',
    },
  };

  override items: CardModel[] = [];

  constructor(options: Record<string, any>) {
    const { name, description, parent } = options;
    super({ name, description, parent });

    this.classList = this.__default.classList;  
  };

  addCard = (newCard: CardModel) => {
    this.addItem(newCard);
  };

  removeCard = (removedCard: CardModel) => {
    this.removeItem(removedCard);
  }
 
};

export default GroupModel;
