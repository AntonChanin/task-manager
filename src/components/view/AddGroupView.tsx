import React, { FC } from 'react';
import { observer } from 'mobx-react-lite';

import Button from '../ui/Button';
import Paper from '../ui/Paper';
import Title from '../ui/Title';
import CardView from './CardView';
import GroupModel from '../../model/GroupModel';
import localization from '../../assets/localization';
import tailwindcssStyles from '../../assets/styles';
import useUpdate from '../../hooks/useUpdate';
import { ViewExtendeble } from '../../types/view';
import { VoidCallback } from '../../types/common';

const AddGroupView: FC<ViewExtendeble<GroupModel, VoidCallback>> = (props) => {
    const { callback, model } = props;
    const {
        name,
        items: cards,
        lang,
        didUpdate,
        makeClass,
    } = model;
    
    useUpdate(didUpdate, []);

    return (
        <Paper className={makeClass(['paper', tailwindcssStyles['theme']['background']['primary']])}>
            {name && <Title value={name} className={makeClass(['title'])} />}
            <div className={makeClass(['cards'])}>
                {cards.map((card) => <CardView key={`card__${card.id}`} model={card} />)}
            </div>
            <Button
                className={makeClass(['button'])}
                theme="dark"
                value={`+ ${localization[lang].addAnotherList}`}
                variant="thirdy"
                callback={callback}
            />
        </Paper>
    );
};

export default observer(AddGroupView);
