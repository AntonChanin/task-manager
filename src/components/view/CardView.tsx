import React, { FC, useState } from 'react';
import { observer } from 'mobx-react-lite';

import Article from '../ui/Article';
import Button from '../ui/Button';
import Paper from '../ui/Paper';
import Textarea from '../ui/Textarea';
import CardModel from '../../model/CardModel';
import InstanceTaskManagerStore from '../../store';
import localization from '../../assets/localization';
import tailwindcssStyles from '../../assets/styles';
import { ViewWithModel } from '../../types/view';

const CardView: FC<ViewWithModel<CardModel>> = (props) => {
    const { model } = props;
    const { id, description, variant, lang, makeClass, setEdit, removeFromParent } = model;
    const { removeCardId } = InstanceTaskManagerStore;
    const [isEdit, setIsEdit] = useState(model.isEdit);

    const cardEdit = () =>  {
        if (model.description.trimEnd()) {
        setIsEdit(!isEdit);
        setEdit(isEdit);
        } else {
        removeFromParent();
        removeCardId(id)
        };
    };

    const updateDescriptionCallback: <T>(props: Record<string, T>) => void =
        (props) => {
            const { value } = props;
            model.description = `${value}`.trim();
        };        

    return (
        <Paper>
            <Button
                className={makeClass(['editButton'])}
                variant={variant}
                callback={cardEdit}
            >
                <img
                width={24}
                height={24}
                src={tailwindcssStyles['icons']['edit']['src']}
                alt={tailwindcssStyles['icons']['edit']['author']}
                />
            </Button>
            {!isEdit
                ? (
                <Article value={description} />
                ) : (
                <Textarea
                    value={description}
                    variant="thirdy"
                    placeholder={localization[lang].editCardPlaceholder}
                    callback={updateDescriptionCallback}
                />
                )
            }
        </Paper>
    );
};

export default observer(CardView);
