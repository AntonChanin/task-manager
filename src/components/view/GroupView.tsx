import { FC, useState } from 'react';
import { observer } from 'mobx-react-lite';

import Button from '../ui/Button';
import Title from '../ui/Title';
import Input from '../ui/Input';
import Paper from '../ui/Paper';
import Row from '../ui/Row';
import CardView from './CardView';
import CardModel from '../../model/CardModel';
import GroupModel from '../../model/GroupModel';
import InstanceTaskManagerStore from '../../store';
import localization from '../../assets/localization';
import tailwindcssStyles from '../../assets/styles';
import useUpdate from '../../hooks/useUpdate';
import { ViewWithModel } from '../../types/view';

const GroupView: FC<ViewWithModel<GroupModel>> = (props) => {
  const { model } = props;
  const {
    id,
    name,
    items: cards,
    lang,
    setEdit,
    addCard,
    didUpdate,
    makeClass,
    removeFromParent,
  } = model;
  const { cardIds, addCardId, removeGroupId } = InstanceTaskManagerStore;
  const [isEdit, setIsEdit] = useState(model.isEdit);

  const cardEdit = () =>  {
    if (model.name.trimEnd()) {
      setIsEdit(!isEdit);
      setEdit(isEdit);
    } else {
      removeFromParent();
      removeGroupId(id);
    }
  };
  
  useUpdate(didUpdate, [cardIds.length, model.name, model.items.length]);

  const addNewCard = () => {
    const card = new CardModel({
      name: '',
      description: '',
      variant: 'thirdy',
      parent: model,
    });
    addCard(card);
    addCardId(card.id);
  };

  const updateNameCallback: <T>(props: Record<string, T>) => void = (props) => {
    const { value } = props;
    model.name = `${value}`;
  }

  return (
    <Paper className={makeClass(['paper', tailwindcssStyles['theme']['background']['thirdy']])}>
      {!isEdit ? (
        <Row>
          <Title value={name} className={makeClass(['title'])} />
          <Button
            variant="thirdy"
            callback={cardEdit}
          >
            <img
              width={24}
              height={24}
              src={tailwindcssStyles['icons']['more']['src']}
              alt={tailwindcssStyles['icons']['more']['author']}
            />
          </Button>
        </Row>
      ) : (
        <Row>
          <Input
            value={name}
            variant="thirdy"
            placeholder={localization[lang].editGroupTitlePlaceholder}
            callback={updateNameCallback}
          />
          <Button
            variant="thirdy"
            callback={cardEdit}
          >
            <img
              width={24}
              height={24}
              src={tailwindcssStyles['icons']['more']['src']}
              alt={tailwindcssStyles['icons']['more']['author']}
            />
          </Button> 
        </Row>
      )}
      <div className={makeClass(['cards'])}>
        {cards.map((card) => <CardView key={`card__${card.id}`} model={card} />)}
      </div>
      <Button
        className={makeClass(['button'])}
        theme="light"
        value={`+ ${localization[lang].addCard}`}
        variant="thirdy"
        callback={addNewCard}
      />
    </Paper>
  );
};

export default observer(GroupView);
