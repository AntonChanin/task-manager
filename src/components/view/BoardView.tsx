import { FC, useEffect } from 'react';
import { observer } from 'mobx-react-lite';

import Title from '../ui/Title';
import GroupModel from '../../model/GroupModel';
import GroupView from './GroupView';
import CardModel from '../../model/CardModel';
import InstanceTaskManagerStore from '../../store';
import localization from '../../assets/localization';
import { BoardProps } from '../../types/view';
import Paper from '../ui/Paper';

const BoardView: FC<BoardProps> = (props) => {
  const { model } = props;
  const { __default: { items: defaultGroup }, classList, items: groups, lang, addGroup } = model;
  const { addGroupId, addCardId, cardIds, groupIds } = InstanceTaskManagerStore;

  useEffect(() => {}, [cardIds.length, groupIds.length])

  const addNewGroup = () => {
    const group = new GroupModel({
      name: 'name',
      description: 'lorem smorem',
    })
    addGroupId(group.id);
    group.addCard(new CardModel({
      name: '',
      description: 'lorem smorem',
    }));
    addCardId(group.items[group.items.length - 1].id);
    addGroup(group);
  };

  return (
    <Paper variant="secondary" className={classList['paper']}>
      <Title className={classList['title']} value={localization[lang].welcomeBoard} />
      <div className={classList['group']}>
        {groups.map((group) => <GroupView model={group}/>)}
        <GroupView callback={addNewGroup} model={defaultGroup[0]}/>
      </div>     
    </Paper>    
  );
};

export default observer(BoardView);
