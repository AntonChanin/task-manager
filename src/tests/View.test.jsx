import React from 'react';
import renderer from 'react-test-renderer';
import {describe, expect, it } from '@jest/globals';

import AddGroupView from '../components/view/AddGroupView';
import BoardView from '../components/view/BoardView';
import CardView from '../components/view/CardView';
import GroupView from '../components/view/GroupView';
import BoardModel from '../model/BoardModel';
import CardModel from '../model/CardModel';
import GroupModel from '../model/GroupModel';
import InstanceTaskManagerStore from '../store';

const { boards } = InstanceTaskManagerStore;

describe('View snapshots:', () => {
    it('AddGroupView renders correctly', () => {
        const tree = renderer
            .create(
                <AddGroupView
                    model={
                        new GroupModel({
                            name: 'name',
                            description: 'lorem smorem',
                            parent: new BoardModel({ name: 'test' }),
                        })
                    }
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('BoardView renders correctly', () => {
        const tree = renderer
            .create(<BoardView model={boards[0]}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('CardView renders correctly', () => {
        const tree = renderer
            .create(
                <CardView
                    model={
                        new CardModel({
                            name: '',
                            description: '',
                            variant: 'thirdy',
                            parent: new GroupModel({
                                name: 'name',
                                description: 'lorem smorem',
                                parent: new BoardModel({ name: 'test' }),
                            }),
                        })
                    }
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('GroupView renders correctly', () => {
        const tree = renderer
            .create(
                <GroupView
                    model={
                        new GroupModel({
                            name: 'name',
                            description: 'lorem smorem',
                            parent: new BoardModel({ name: 'test' }),
                        })
                    }
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});