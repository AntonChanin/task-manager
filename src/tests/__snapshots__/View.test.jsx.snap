// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View snapshots: AddGroupView renders correctly 1`] = `
<div
  className="w-1/4 min-w-[25%] h-max px-4 mr-2 flex-col items-start bg-blue-500 border-radius border-none rounded-xl p-2 mb-4 text-start"
>
  <div
    className="font-sans font-bold flex place-content-between py-4"
  >
    name
  </div>
  <div
    className="overflow-y-auto max-h-[78vh] pr-2"
  />
  <button
    className="text-white border-none outline-none m-2 hover: text-gray-100
     flex"
    onClick={[Function]}
  >
    + Add another list
  </button>
</div>
`;

exports[`View snapshots: BoardView renders correctly 1`] = `
<div
  className="flex-col bg-blue-400 p-4 font-sans w-full h-full border-none p-2 mb-4 text-start"
>
  <div
    className="font-sans font-bold my-2 w-max text-white text-2xl"
  >
    Welcome board
  </div>
  <div
    className="flex max-w-full overflow-x-auto overflow-y-hidden h-[95%]"
  >
    <div
      className="w-1/4 min-w-[25%] h-max px-4 mr-2 flex-col items-start bg-blue-500 border-radius border-none rounded-xl p-2 mb-4 text-start"
    >
      <div
        className="overflow-y-auto max-h-[78vh] pr-2"
      />
      <button
        className="text-white border-none outline-none m-2 hover: text-gray-100
     flex"
        onClick={[Function]}
      >
        + Add another list
      </button>
    </div>
  </div>
</div>
`;

exports[`View snapshots: CardView renders correctly 1`] = `
<div
  className="drop-shadow-md bg-white border-radius border-none rounded-xl p-2 mb-4 text-start"
>
  <button
    className="text-gray-400 border-none outline-none m-2 hover: text-gray-600
     opacity-0 flex ml-auto hover:opacity-100 absolute top-0 right-0"
    onClick={[Function]}
  >
    <img
      alt="Draw icons created by Gregor Cresnar - Flaticon"
      height={24}
      src="src/assets/icons/edit.png"
      width={24}
    />
  </button>
  <textarea
    className="w-full outline-none border-none m-auto"
    onChange={[Function]}
    placeholder="Add new card description"
    value=""
  />
</div>
`;

exports[`View snapshots: GroupView renders correctly 1`] = `
<div
  className="w-1/4 min-w-[25%] h-max px-4 mr-2 flex-col items-start bg-gray-100 border-radius border-none rounded-xl p-2 mb-4 text-start"
>
  <div
    className="flex justify-between"
  >
    <input
      className="w-11/12 w-full outline-none border-none m-auto"
      onChange={[Function]}
      placeholder="Add new list title"
      value="name"
    />
    <button
      className="text-gray-400 border-none outline-none m-2 hover: text-gray-600"
      onClick={[Function]}
    >
      <img
        alt="View more icons created by Any Icon - Flaticon"
        height={24}
        src="src/assets/icons/more.png"
        width={24}
      />
    </button>
  </div>
  <div
    className="overflow-y-auto max-h-[78vh] pr-2"
  />
  <button
    className="text-gray-400 border-none outline-none m-2 hover: text-gray-600
     flex"
    onClick={[Function]}
  >
    + Add another card
  </button>
</div>
`;
